bundle-workflow/python/test_workflow/test_cluster.py

```diff
@@ -1,52 +1,41 @@
-import os
-import tempfile
-import urllib.request
-import shutil
-import subprocess
+import abc
 
-class LocalTestCluster:
-    def __init__(self, bundle_manifest):
-        self.manifest = bundle_manifest
-        self.work_dir = tempfile.TemporaryDirectory()
+class TestCluster(abc.ABC):
+    '''
+    Abstract base class for all types of test clusters.
+    '''
 
+    @abc.abstractmethod
     def create(self):
-        print(f'Creating local test cluster in {self.work_dir.name}')
-        os.chdir(self.work_dir.name)
-        print(f'Downloading bundle from {self.manifest.build.location}')
-        urllib.request.urlretrieve(self.manifest.build.location, 'bundle.tgz')
-        print(f'Downloaded bundle to {os.path.realpath("bundle.tgz")}')
-        print('Unpacking')
-        subprocess.check_call('tar -xzf bundle.tgz', shell = True)
-        print('Unpacked')
-        self.stdout = open('stdout.txt', 'w')
-        self.stderr = open('stderr.txt', 'w')
-        dir = f'opensearch-{self.manifest.build.version}'
-        self.process = subprocess.Popen('./opensearch-tar-install.sh', cwd = dir, shell = True, stdout = self.stdout, stderr = self.stderr)
-        print(f'Started OpenSearch with PID {self.process.pid}')
+        '''
+        Set up the cluster. When this method returns, the cluster must be available to take requests.
+        Throws ClusterCreationException if the cluster could not start for some reason. If this exception is thrown, the caller does not need to call "destroy".
+        '''
+        pass
 
+    @abc.abstractmethod
+    def destroy(self):
+        '''
+        Tear down the cluster. If the cluster is already destroyed or has not yet been created then this is a no-op.
+        '''
+        pass
+
+    @abc.abstractmethod
     def endpoint(self):
-        return 'localhost'
+        '''
+        Get the endpoint that this cluster is listening on, e.g. 'localhost' or 'some.ip.address'.
+        '''
+        pass
 
+    @abc.abstractmethod
     def port(self):
-        return 9200
+        '''
+        Get the port that this cluster is listening on.
+        '''
+        pass
 
-    def destroy(self):
-        print(f'Sending SIGTERM to PID {self.process.pid}')
-        self.process.terminate()
-        try:
-            print('Waiting for process to terminate')
-            self.process.wait(10)
-        except TimeoutExpired:
-            print('Process did not terminate after 10 seconds. Sending SIGKILL')
-            self.process.kill()
-            try:
-                print('Waiting for process to terminate')
-                self.process.wait(10)
-            except TimeoutExpired:
-                print('Process failed to terminate even after SIGKILL')
-                raise
-        finally:
-            print(f'Process terminated with exit code {self.process.returncode}')
-            self.stdout.close()
-            self.stderr.close()
-            self.process = None
+class ClusterCreationException(Exception):
```

Comment:
nit: Can't we use this class to throw error while creating the cluster?

Reply:
That's the idea, yeah. I'll update LocalTestCluster to throw this exception when it fails.